src/webgpu/api/operation/render_pipeline/pipeline_output_targets.spec.ts

```diff
@@ -1,12 +1,399 @@
+  getExpectedTypeAndComponentCount(
+    format: GPUTextureFormat
+  ): { expectedType: GPUTextureSampleType; expectedComponentCount: number } {
+    let expectedType: GPUTextureSampleType;
+    if (format.endsWith('sint')) {
+      expectedType = 'sint';
+    } else if (format.endsWith('uint')) {
+      expectedType = 'uint';
+    } else {
+      expectedType = 'float';
+    }
+    // Only used for renderable color formats
+    let expectedComponentCount: number = 1;
+    if (format.startsWith('rgba') || format.startsWith('bgra') || format.startsWith('rgb10a2')) {
+      expectedComponentCount = 4;
+    } else if (format.startsWith('rg')) {
+      expectedComponentCount = 2;
+    } else if (format.startsWith('r')) {
+      expectedComponentCount = 1;
+    } else {
+      unreachable();
+    }
+
+    return { expectedType, expectedComponentCount };
+  }
+
+  getFragmentShaderCode(
+    output: readonly number[],
+    sampleType: GPUTextureSampleType,
+    componentCount: number
+  ): string {
+    let fragColorType;
+    let suffix;
+    let fractionDigits = 0;
+    switch (sampleType) {
+      case 'sint':
+        fragColorType = 'i32';
+        suffix = '';
+        break;
+      case 'uint':
+        fragColorType = 'u32';
+        suffix = 'u';
+        break;
+      default:
+        fragColorType = 'f32';
+        suffix = '';
+        fractionDigits = 2;
+        break;
+    }
+
+    const v = output.map(n => n.toFixed(fractionDigits));
+
+    let outputType;
+    let result;
+    switch (componentCount) {
+      case 1:
+        outputType = fragColorType;
+        result = `${v[0]}${suffix}`;
+        break;
+      case 2:
+        outputType = `vec2<${fragColorType}>`;
+        result = `${outputType}(${v[0]}${suffix}, ${v[1]}${suffix})`;
+        break;
+      case 3:
+        outputType = `vec3<${fragColorType}>`;
+        result = `${outputType}(${v[0]}${suffix}, ${v[1]}${suffix}, ${v[2]}${suffix})`;
+        break;
+      case 4:
+        outputType = `vec4<${fragColorType}>`;
+        result = `${outputType}(${v[0]}${suffix}, ${v[1]}${suffix}, ${v[2]}${suffix}, ${v[3]}${suffix})`;
+        break;
+      default:
+        unreachable();
+    }
+
+    return `
+    [[stage(fragment)]] fn main() -> [[location(0)]] ${outputType} {
+        return ${result};
+    }`;
+  }
+}
+
+export const g = makeTestGroup(F);
+
+g.test('color,component_count')
+  .desc(
+    `Test that extra components of the output (e.g. f32, vec2<f32>, vec3<f32>, vec4<f32>) are discarded.`
+  )
+  .params(u =>
+    u
+      .combine('format', kRenderableColorTextureFormats)
+      .beginSubcases()
+      .combine('componentCount', [1, 2, 3, 4])
+  )
+  .fn(async t => {
+    const { format, componentCount } = t.params;
+    const { expectedType, expectedComponentCount } = t.getExpectedTypeAndComponentCount(format);
+    if (componentCount < expectedComponentCount) {
+      t.skip('componentCount of pipeline output must not be fewer than that of target format');
```

Comment:
For anything that can be "statically" filtered (without having a webgpu device), prefer to use `.unless` or `.filter` in the params instead.